apm-agent-core/src/main/java/co/elastic/apm/agent/impl/CloudMetadataProvider.java

```diff
@@ -0,0 +1,401 @@
+                        cloudProviderInfo = gcpMetadata.get(futureTimeout, TimeUnit.MILLISECONDS);
+                    }
+                    break;
+                }
+            }
+        } catch (Throwable throwable) {
+            unexpectedError = throwable;
+        }
+        logSummary(cloudProvider, cloudProviderInfo, unexpectedError);
+        return cloudProviderInfo;
+    }
+
+    private static void logSummary(CoreConfiguration.CloudProvider cloudProvider,
+                                   @Nullable CloudProviderInfo cloudProviderInfo, @Nullable Throwable unexpectedError) {
+        if (cloudProviderInfo == null) {
+            if (cloudProvider == AWS || cloudProvider == AZURE || cloudProvider == GCP) {
+                String msg = cloudProvider.name() + " is defined as the cloud_provider, but no metadata was found where expected";
+                if (unexpectedError != null) {
+                    logger.warn(msg, unexpectedError);
+                } else {
+                    logger.warn(msg);
+                }
+            } else if (unexpectedError != null) {
+                logger.warn("Unexpected error during automatic discovery process for cloud provider", unexpectedError);
+            } else {
+                logger.debug("cloud_provider configured {}, no cloud metadata discovered", cloudProvider);
+            }
+        } else {
+            logger.debug("Cloud metadata discovered: {}", cloudProviderInfo);
+        }
+    }
+
+    @Nullable
+    private static CloudProviderInfo getAwsMetadata(int queryTimeoutMs, CoreConfiguration.CloudProvider configuredProvider) throws IOException {
+        String awsTokenUrl = "http://169.254.169.254/latest/api/token";
+        Map<String, String> headers = new HashMap<>(1);
+        headers.put("X-aws-ec2-metadata-token-ttl-seconds", "300");
+        String token = null;
+        try {
+            token = executeRequest(awsTokenUrl, "PUT", headers, queryTimeoutMs);
+            logger.debug("Got aws token with a length of {} characters", token.length());
+        } catch (Exception e) {
+            if (configuredProvider == AWS) {
+                // This is expected when the token request is made from within a Docker container as described in https://github.com/elastic/apm-agent-python/pull/884
+                logger.info("Unable to obtain API token, probably because running within a Docker container. This means that AWS metadata may not be available.");
+            }
+        }
+        String awsMetadataUrl = "http://169.254.169.254/latest/dynamic/instance-identity/document";
+        Map<String, String> documentHeaders = null;
+        if (token != null) {
+            documentHeaders = new HashMap<>(1);
+            documentHeaders.put("X-aws-ec2-metadata-token", token);
+        }
+        String metadata = executeRequest(awsMetadataUrl, "GET", documentHeaders, queryTimeoutMs);
+        logger.debug("AWS metadata retrieved");
```

Comment:
maybe output JSON for easy debugging

Reply:
I did that and got very big output, including SSH keys 😬 ...